services/src/main/java/org/keycloak/protocol/saml/JaxrsSAML2BindingBuilder.java

```diff
@@ -33,25 +37,48 @@
  * @version $Revision: 1 $
  */
 public class JaxrsSAML2BindingBuilder extends BaseSAML2BindingBuilder<JaxrsSAML2BindingBuilder> {
-    public static class PostBindingBuilder extends BasePostBindingBuilder {
+
+    private final KeycloakSession session;
+
+    public JaxrsSAML2BindingBuilder(KeycloakSession session) {
+        this.session = session;
+    }
+
+    public class PostBindingBuilder extends BasePostBindingBuilder {
         public PostBindingBuilder(JaxrsSAML2BindingBuilder builder, Document document) throws ProcessingException {
             super(builder, document);
         }
+
         public Response request(String actionUrl) throws ConfigurationException, ProcessingException, IOException {
-            return buildResponse(document, actionUrl, true);
+            return createResponse(actionUrl, GeneralConstants.SAML_REQUEST_KEY);
         }
+
         public Response response(String actionUrl) throws ConfigurationException, ProcessingException, IOException {
-            return buildResponse(document, actionUrl, false);
+            return createResponse(actionUrl, GeneralConstants.SAML_RESPONSE_KEY);
         }
-        protected Response buildResponse(Document responseDoc, String actionUrl, boolean asRequest) throws ProcessingException, ConfigurationException, IOException {
-            String str = builder.buildHtmlPostResponse(responseDoc, actionUrl, asRequest);
 
-            return Response.ok(str, MediaType.TEXT_HTML_TYPE)
-                    .header("Pragma", "no-cache")
-                    .header("Cache-Control", "no-cache, no-store").build();
+        private Response createResponse(String actionUrl, String key) throws ProcessingException, ConfigurationException, IOException {
+            MultivaluedMap<String,String> formData = new MultivaluedHashMap<>();
+            formData.add(GeneralConstants.URL, actionUrl);
+
+            switch (key) {
+                case GeneralConstants.SAML_REQUEST_KEY:
+                    formData.add(GeneralConstants.SAML_REQUEST_KEY, builder.getSAMLResponse(document));
```

Comment:
nit: replacing with
```suggestion
                    formData.add(key, builder.getSAMLResponse(document));
```
would allow also handling `GeneralConstants.SAML_RESPONSE_KEY` in one case block.